### PR TITLE
Support Custom Memory Plan in CudaCachingAllocator

### PR DESCRIPTION
**Pull request by o-hanna**
We modify CudaCachingAllocator to support a custom memory plan given from the python side. This can be done by calling the function "torch.cuda.set_memory_plan" with inputs:
- plan: list of memory plans for each device. Each plan is a list of AllocFreeEvents (see PR: https://github.com/pytorch/pytorch/pull/81753) with member variables size (expected allocation size) and ptr (offset for allocation).
- plan_size: list of the size needed for plan execution for each device.
- non_rec_length (optional): the plan is expected to be repeated after the last allocation. If there is transient events in the beginning, that are not repeated, the number of those events should be passed to "non_rec_length[device]", which is 0 by default.

Allocation: For each device CudaCachingAllocator will match the size of allocation with the expected allocation size given in the plan; if they match, the event will be allocated according to the plan; otherwise, the event will be passed to the normal allocator.
Free: CudaCachingAllocator checks if the event is allocated according to the plan or the normal allocator, then call the appropriate free function.

Now each free/allocate event includes the following bool flags:

raw_alloc: indicates if the allocation is raw
alloc_type: enum {served_by_new_block, \\ indicates if the allocation is served by a new block
served_by_new_block_retry, \\ indicates if the allocation is served by a new block after freeing one cached block
defrag \\ indicates if the allocation is served by a new block after defragmentation (freeing all cached blocks)}

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:498]: curious, what are the reasons to use `ska::flat_has_map` here instead of `std::unordered_map`?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:500]: looks like this can be const reference?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:515]: please add comments to explain the behavior of this method

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:519]: hmm, I thought `size` is `size_t` type. Any reason for converting RHS to `unsigned int` before comparison?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:518]: is this line necessary? This needs to be the same as `size`. So can `size` be zero or negative integer?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:528]: unnecessary new lines?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:529]: could you please clang-format the code before submitting for review?

File: torch/cuda/__init__.py
Comment (mrshenli) [inline torch/cuda/__init__.py:370]: Could you please check if these can render properly?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:2076]: since you are not writing plan in this method, `plan` can be a const reference to avoid copy?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:2086]: > failed to assign enough memory for planning

this will throw an OOM error as usual? Please add tests for it.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:2096]: if you want this to be a singleton, you can borrow how `RRefContext::getInstance()` was implemented.

File: torch/cuda/__init__.py
Comment (mrshenli) [inline torch/cuda/__init__.py:380]: this is not tested at all?

File: c10/cuda/CUDACachingAllocator.h
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.h:129]: hmm, this can be an enum?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:703]: Do you plan to allow users to toggle the new algorithm instead of forcing it for all allocations?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:547]: I guess `block == nullptr` means planned allocation failed? If so, please add comments to highlight that.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:529]: `planned_ptr+ s ize`

how does above even pass build?

**Review by mrshenli — CHANGES_REQUESTED** [review]

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:529]: Sorry about that. It build successfully but I trusted "lintrunner -a" after build. It caused a lot of issues. I am going through the code again and will update the PR.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:519]: We use the same event struct used for tracking. The size is an int as we use negative to indicate free. I convert it to suppress comparison warning.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:495]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:515]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:547]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:703]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:500]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:2086]: I changed this part. Thank you.

File: c10/cuda/CUDACachingAllocator.h
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.h:129]: Done. Thank you.

File: torch/cuda/__init__.py
Comment (o-hanna) [reply torch/cuda/__init__.py:380]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:565]: I assume you need your mutex to protect these?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [reply c10/cuda/CUDACachingAllocator.cpp:502]: Oh I see, you mean non-recurring

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:567]: this means we move forward regardless of the current allocation is expected or not?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:577]: do you need mutex?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:598]: would I be correct that we should only append when this is active and error out otherwise?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:607]: `emplace_back`?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:600]: which initialization are you referring? you mean the `resize(1)` below?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:701]: try to avoid exposing member vars as public if not necessary. For this one, you can expose an `is_active` member method.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:706]: this is not planned, do we need to print a warning?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:707]: This is OK for now to unblock you, but it might be more reasonable to keep existing allocator intact and make the new allocator the entry point within the tracking context. And the new allocator can call into the old one if it fails o allocate blocks.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (mrshenli) [inline c10/cuda/CUDACachingAllocator.cpp:1777]: why do you need `resize(1)` here?

File: docs/source/cuda.rst
Comment (mrshenli) [inline docs/source/cuda.rst:114]: in follow PRs, creating a context manager for it.

File: test/test_cuda.py
Comment (mrshenli) [inline test/test_cuda.py:321]: curious, why set/disable instead of enable/disable?

**Review by mrshenli — APPROVED** [review]
This version looks much better. Thanks @o-hanna! I stamped to unblock you so that you can move fast, but please do address comments before landing.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:502]: Yes non-recurring :)

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:508]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:510]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:548]: This is just to be extra careful that the memory plan is feasible since it is passed from the python side and we do not have control over it. If some allocations are not feasible according to plan, we pass those allocations to the normal CudaCachingAllocator.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:565]: Since this memory_planner.malloc is only called from inside device_allocator.malloc, then it is already protected?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:577]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:567]: Since this is inside "if ((plan_blk_ptr[device] != nullptr) && (size == (uint32_t)mem_plan[device][plan_counter[device]].size))", we only move if the allocation is expected? However, if the allocation is expected but not allocated according to plan, we move forward. The later case can happen in two scenarios:
1. Plan is not feasible.
2.  Unexpected event with size and order that matches the expected event is allocated according to plan but freed later than expected by plan.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:523]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:525]: Done. Thank you.

File: docs/source/cuda.rst
Comment (o-hanna) [reply docs/source/cuda.rst:114]: Sure. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:706]: Do we expect the allocations to be perfectly periodic? If not could it be annoying to keep printing warning?

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:598]: If it is not active, we just do not append.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:600]: Yes, the resize(1). I changed it.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:1777]: I removed it and edited the append function. Thank you.

File: test/test_cuda.py
Comment (o-hanna) [reply test/test_cuda.py:321]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp:701]: Done. Thank you.

File: c10/cuda/CUDACachingAllocator.cpp
Comment (o-hanna) [reply c10/cuda/CUDACachingAllocator.cpp]: Done. Thank you.

**Pull request merged**
